tests/Model/LazyBSONArrayTest.php

```diff
@@ -0,0 +1,289 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use Generator;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Exception\InvalidArgumentException;
+use MongoDB\Model\LazyBSONArray;
+use MongoDB\Model\LazyBSONDocument;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+use function iterator_to_array;
+use function json_encode;
+use function serialize;
+use function unserialize;
+
+use const JSON_THROW_ON_ERROR;
+
+class LazyBSONArrayTest extends TestCase
+{
+    public static function provideTestArray(): Generator
+    {
+        yield 'array' => [
+            new LazyBSONArray([
+                'bar',
+                new LazyBSONDocument(['bar' => 'baz']),
+                new LazyBSONArray([0, 1, 2]),
+            ]),
+        ];
+
+        yield 'packedArray' => [
+            new LazyBSONArray(PackedArray::fromPHP([
```

Comment:
Why does the `array` test case above wrap nested arrays/documents with lazy objects while this does not?

Reply:
This is a side effect of BSOn decoding. Since anything we're reading from BSON is passed through the codec library, nested `PackedArray` and `Document` instances are converted to their lazy counterparts. When creating a `LazyBSONArray` from PHP data, the data is taken as-is. This behaviour is consistent with how `BSONArray` operates, where nested arrays are also not converted to `BSONArray` instances.

I can add a comment explaining this, but in this case I decided to create the lazy instances in the data provider so that the tests can be reused for both data sets.